# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Moezulhaq24/FlyRank-Internship-ML/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

I am choosing **Lane 2: Refresh / Content Opportunity Scoring**

This lane asks: which pages should be reviewed first for refresh, expansion, CTR review, engagement improvement, or monitoring? I chose this lane because the starter notebooks already showed that page-level signals like impressions, trend direction, CTR, average position, content age, and engagement can help prioritize content review. This is useful because content teams have limited time, so the goal is not to automatically change pages, but to support human reviewers with a ranked list of pages and clear reason codes.

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

My research question is:

**Which existing content pages should be reviewed first for refresh, expansion, CTR review, engagement improvement, or monitoring?**

The decision this work improves is content prioritization: deciding which pages deserve human review first. The people who act on this output are SEO analysts, content editors, or content strategists. The action they can take is to refresh old content, expand thin content, review title/meta for CTR, improve engagement, or monitor low-confidence pages.

A wrong recommendation can waste editor time, push a low-value page above a more important page, or cause an important declining page to be missed. Because of this, the output should not be only a score. It should include reason codes and confidence labels so a human can understand why a page was recommended.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

The starter dataset gives enough real signals to justify this lane. I will check how many pages exist, how many are declining, how many declining pages still have search demand, and how many visible pages have low CTR. These numbers help show why a refresh-prioritization queue can be useful.

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import os
import subprocess
import pandas as pd
import numpy as np

DATA_PATH = "data/raw/content_refresh_anonymized.csv"

# If running in Colab and the data file is not available, clone my repo.
if not os.path.exists(DATA_PATH):
    REPO_URL = "https://github.com/Moezulhaq24/FlyRank-Internship-ML"
    REPO_DIR = "FlyRank-Internship-ML"

    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)

    os.chdir(REPO_DIR)

df = pd.read_csv(DATA_PATH)

declining = df["trend_direction"].str.lower().eq("down")
visible_pages = df["impressions_90d"] >= 500
declining_with_demand = declining & (df["impressions_90d"] >= 100)

low_ctr_visible = (
    (df["impressions_90d"] >= 500)
    & (df["avg_position"] > 0)
    & (df["avg_position"] <= 20)
    & (df["ctr"] < 0.5)
)

low_engagement_visible = (
    (df["sessions_90d"] >= 30)
    & (
        ((df["engagement_rate"] > 0) & (df["engagement_rate"] < 30))
        | ((df["scroll_rate"] > 0) & (df["scroll_rate"] < 30))
    )
)

search_volume_corr = df["search_volume"].corr(df["impressions_90d"])

print("Total pages:", len(df))
print("Total columns:", df.shape[1])
print("Total pseudonymized clients:", df["client_id"].nunique())

print("\nDeclining pages:", declining.sum())
print("Declining rate:", round(declining.mean(), 3))

print("\nVisible pages with impressions_90d >= 500:", visible_pages.sum())
print("Declining pages with impressions_90d >= 100:", declining_with_demand.sum())
print("Visible pages with low CTR:", low_ctr_visible.sum())
print("Visible pages with weak engagement:", low_engagement_visible.sum())

print("\nCorrelation between search_volume and impressions_90d:", round(search_volume_corr, 4))

Total pages: 30000
Total columns: 44
Total pseudonymized clients: 32

Declining pages: 16262
Declining rate: 0.542

Visible pages with impressions_90d >= 500: 16726
Declining pages with impressions_90d >= 100: 13152
Visible pages with low CTR: 9759
Visible pages with weak engagement: 6508

Correlation between search_volume and impressions_90d: 0.0012


##Result Explanation

These numbers support my lane choice. The starter dataset contains 30,000 pages across 32 pseudonymized clients. Around 54.2% of pages are marked as declining in the starter label, which means there is enough decline signal to study.

There are also many pages where action could matter: 13,152 declining pages still have at least 100 impressions, and 9,759 visible pages have low CTR. This suggests that content review should not only look at whether a page is old or declining. It should also consider demand, CTR, position, engagement, and freshness together.

The correlation between search volume and actual page impressions is almost zero, so keyword volume alone is not enough to decide which pages deserve review first. This supports using a data-driven ranking approach instead of one simple rule.

## 4. Careful words: what I can and can't claim

This work can make observed, directional, decision-support claims. I can say that some pages show measured signals such as decline, visibility, low CTR, weak engagement, age, or freshness risk. I can also say that a ranked queue may help editors decide which pages to review first.

I cannot claim that a recommended page will definitely recover traffic after refresh. I also cannot claim that this predicts Google's algorithm or proves that any signal causes ranking changes. The output should be treated as a human review aid, not an automatic publishing decision.

For this lane, the safest claim is: the data can help prioritize review candidates, but final content decisions still need human judgment.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [X] The notebook runs top to bottom with no errors (Runtime → Run all)
- [X] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.